src/BoundSmallAllocations.cpp

```diff
@@ -14,16 +14,38 @@ class BoundSmallAllocations : public IRMutator {
     // Track constant bounds
     Scope<Interval> scope;
 
+    template<typename T, typename Body>
+    Body visit_let(const T *op) {
```

Comment:
This is the best example of the idiom I use throughout. Make a vector of frames on the heap storing the state that used to be on the stack, build it up using a do while loop, mutate the body, then unwind by reverse iterating over the frames.

Reply:
Is the idiom regular enough that it could be template-ized somehow?

Reply:
I thought about that for a while, and I don't believe so. There might be something clever with lambdas to capture state or something, but it escaped me.

Reply:
What about this (for Block, the same thing could work for Let/LetStmt):

- Add visitor methods `IRMutator(vector<const Block*>)` and `IRVisitor::visit(vector<const Block*>)`
- Implement IRMutator::visit(const Block*) and IRVisitor::visit(const Block*) to flatten Blocks into vectors of Blocks, call the visitors above, and then expand them back into IR Blocks.

This probably doesn't work due to C++ overloading/visitor pattern stuff, but it seems like it *might* work. Maybe it should be in IRVisitor/IRMutator derived classes to avoid the overhead when it isn't necessary.

Reply:
It's not the data structure that's the problem. The visitor for consuming a linked list of Blocks looks the same as the visitor for consuming a vector of Blocks - all you need to do is forward-iterate over it. Pre-packing things into a vector just adds overhead.